Backup

sim6 opened the pull request.

NightTsarina commented on playbooks/backup.yaml:
```diff
@@ -0,0 +1,48 @@
+        group: root
+        mode: '0700'
+
+    - name: Retrieve restic password
+      delegate_to: localhost
+      ansible.builtin.command:
+        cmd: ../scripts/pass {{ inventory_hostname }}/restic
+        chdir: '{{ lookup("env", "PWD") }}'
+      become: false
+      register: backup__restic_pass
+      changed_when: false
+
+    - name: Save restic password
+      ansible.builtin.copy:
+        content: '{{ backup__restic_pass.stdout }}'
```
You don't need to execute pass manually, you can use the passwordstore lookup like this: `{{ lookup('community.general.passwordstore', inventory_hostname + '/restic') }}`. But also, both the details of where and how passwords are stored should be left for the caller to decide, in here you can assume you get the cleartext password in a variable

```suggestion
    - name: Save restic password
      ansible.builtin.copy:
        content: '{{ backup__restic_password }}'
```

sim6 replied:
Good to know about the `community.general.passwordstore`. Thanks!

NightTsarina commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,150 @@
+      fi
+    done
+  else
+    if ! mount -m -o ro /dev/$volume-snap $BACKUP_FROM_DIR/$volume
```
why not use `is_filesystem` here?

sim6 replied:
We can use it and it will work now, but the file system detection should be improved because if there is a swap partition, it is detected incorrectly. I'm going to document it.

sim6 replied:
Done via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/1d0df128f8efb47628027e8ba61d39b610af117a

sim6 replied:
I improved filesystem detection via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/00e5a224b88e3bc9469c6ac794c0fda7392fdc1b

NightTsarina commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,150 @@
+  echo "ERROR: Environment variable RESTIC_PASSWORD, RESTIC_PASSWORD_COMMAND or RESIC_PASSWORD_FILE should be defined" > /dev/stderr &&
+    exit 1
+
+trap teardown HUP INT QUIT ABRT ALRM TERM KILL
```
you can just trap "EXIT" instead so the function is called regardless of how the script ended (plus, you cannot trap SIGKILL, as the kernel kills the process immediately)
```suggestion
trap teardown EXIT
```

sim6 replied:
Thanks!

sim6 replied:
Should we add INT to catch Control-C when running on interactive console?

NightTsarina commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,150 @@
+done
+
+restic backup $BACKUP_FROM_DIR
+
+teardown
```
if you go ahead with trapping `EXIT`, then this is not needed
```suggestion
```

sim6 replied:
Thanks!

sim6 commented:
`kpartx -l` is not working in snapshots of virtual machine root volume of iris.infra.hivecompute.ai.

Commands:
```
> ssh -t iris.infra.hivecompute.ai sudo lvcreate -
-snapshot -L 1g -n vm-cosmos-root-snap /dev/vg0/vm-cosmos-root
[sudo] password for sim6:
  Logical volume "vm-cosmos-root-snap" created.
Connection to iris.infra.hivecompute.ai closed.
> ssh -t iris.infra.hivecompute.ai sudo partx -l /dev/vg0/vm-cosmos-root-snap
[sudo] password for sim6:
Connection to iris.infra.hivecompute.ai closed.
```
No partitions detected.

Clean up:
```
> ssh -t iris.infra.hivecompute.ai sudo lvremove /dev/vg0/vm-cosmos-root-snap
[sudo] password for sim6:
Do you really want to remove active logical volume vg0/vm-cosmos-root-snap? [y/n]: y
  Logical volume "vm-cosmos-root-snap" successfully removed.
Connection to iris.infra.hivecompute.ai closed.
```

NightTsarina reviewed (approved):
I haven't tested it, but it looks great! Just a couple of small comments, not blocking

PS: we will need some info on how to use restic now :)

NightTsarina commented:
> `kpartx -l` is not working in snapshots of virtual machine root volume of iris.infra.hivecompute.ai.

This is weird.. The  only thing I find that could explain it, is that `fdisk` reports some errors:

```
$ sudo fdisk -l /dev/vg0/vm-cosmos-root > /dev/null
GPT PMBR size mismatch (20971519 != 2621439) will be corrected by write.
Partition 1 does not start on physical sector boundary.
```

sim6 commented:
I squashed the commits.

sim6 commented:
I rebased onto `main`.

sim6 commented:
I added checks for `backup__restic_password`.

dato commented on playbooks/backup.yaml:
```diff
@@ -0,0 +1,42 @@
+        dest: /usr/local/etc/restic_password
+        owner: root
+        group: root
+        mode: '0700'
```
```suggestion
        mode: '0600'
```

sim6 replied:
Ops, thanks!

dato commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,150 @@
+      do
+        findmnt /dev/mapper/$loop_device -o TARGET | while read mount_point
+        do
+	  umount /dev/mapper/$loop_device
+	  rmdir $mount_point
+	done
+	test -f $(get_backup_location $loop_device) &&
```
These lines (and several others below) are indented with tabs, resulting in wrong alignment depending on configured tab width. Could you perhaps ensure all lines are indented with spaces?

sim6 replied:
Apologies, I will fix it!

sim6 replied:
Done via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/6070b90f5b02e2b9eb1b856404ece16b864f6503

dato commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,150 @@
+      kpartx -d /dev/$volume-snap
+      rm -f $BACKUP_FROM_DIR/$volume.mbr
+      rm -f $BACKUP_FROM_DIR/$volume.sfdisk
+    else
+      mountpoint -q $BACKUP_FROM_DIR/$volume &&
+        umount $BACKUP_FROM_DIR/$volume &&
+	  rmdir $BACKUP_FROM_DIR/$volume
+      test -f $BACKUP_FROM_DIR/$volume &&
+        rm $BACKUP_FROM_DIR/$volume
```
This is difficult to parse for correctness without realizing `mountpoint / test -f` are mutually exclusive. How about signalling that with an `else if`?

```suggestion
    elif mountpoint -q $BACKUP_FROM_DIR/$volume
    then
      umount $BACKUP_FROM_DIR/$volume &&
        rmdir $BACKUP_FROM_DIR/$volume
    elif test -f $BACKUP_FROM_DIR/$volume
    then
      rm $BACKUP_FROM_DIR/$volume
```

sim6 replied:
Thanks!

dato commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,150 @@
+        dd if=/dev/mapper/$loop_device of=$(get_backup_location $loop_device)
+      fi
+    done
+  else
+    if is_filesystem /dev/$volume-snap
+    then
+      mount -m -o ro /dev/$volume-snap $BACKUP_FROM_DIR/$volume
+    else
+      dd if=/dev/$volume-snap of=$BACKUP_FROM_DIR/$volume
+    fi
```
```suggestion
  elif is_filesystem /dev/$volume-snap
  then
    mount -m -o ro /dev/$volume-snap $BACKUP_FROM_DIR/$volume
  else
    dd if=/dev/$volume-snap of=$BACKUP_FROM_DIR/$volume
```

sim6 replied:
Thanks!

sim6 commented:
I squashed the commits.

sim6 commented:
I impoved file system detection by checking if it is supported.

sim6 commented:
I improved a comment https://github.com/Tina-pm/ansible-playbooks/compare/03bca9a429b213527911155d1a230e69dbd4b4e8..a1b168d4086c7576e8822d35d3e99e342ea71639.

sim6 commented:
I fixed and improved the backup script via https://github.com/Tina-pm/ansible-playbooks/compare/a1b168d4086c7576e8822d35d3e99e342ea71639..2dbd885d002b370a057a73f8fc53eb89f1f6105b.

sim6 commented:
Another small improvement via https://github.com/Tina-pm/ansible-playbooks/compare/2dbd885d002b370a057a73f8fc53eb89f1f6105b..d7afef6bd33f8daf9fd288c769910a05e3c23f4c.

dato commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,165 @@
+is_filesystem(){
+  # Exit with 0 if the parameter is a device with a supported filesystem, 1 otherwise.
+  local device=$1
+  grep -q "\b$(blkid $device | sed 's/.* TYPE="\([^"]*\)".*/\1/')$" /proc/filesystems
```
```suggestion
  grep -q "\b$(blkid -s TYPE -o value $device)$" /proc/filesystems
```

sim6 replied:
Oh nice!

sim6 commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,165 @@
+is_partition_table(){
+  # Exit with 0 if the parameter is a device with a partition table, 1 otherwise.
+  local device=$1
+  blkid $1 | grep -q " PTTYPE="
```
Ops, thanks!

sim6 commented:
Like @dato suggested for filesystem detection via https://github.com/Tina-pm/ansible-playbooks/pull/20#discussion_r1893982416 I changed partition table detection to use the flags `-s` and `-o` of `blkid` instead of parse the output. 
I also updated the copyright, rebased and squashed the commits.

NightTsarina commented on playbooks/backup.yaml:
```diff
@@ -0,0 +1,42 @@
+      when: backup__restic_password is defined
+
+    - name: Set up cron job
+      ansible.builtin.cron:
```
At first this looked fine, but then I looked at the documentation of the `cron` module, and I think we should avoid it! It primarily manages per-user cron tabs (as opposed to system cron tabs stored in /etc/cron*), and has some really weird ideas on how to do things...

Instead, I'd suggest just using `copy` or `template` to place a file in `/etc/cron.daily`. You could also use `/etc/cron.d`, but I am not sure we need to customise the time of the backup..

sim6 replied:
I done it via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/666a58dc739e6f8202f75ad934e854f0ec4c544c

NightTsarina commented on playbooks/backup.yaml:
```diff
@@ -0,0 +1,42 @@
+          RESTIC_REPOSITORY={{ backup__restic_repository }}
+          /usr/local/sbin/backup.sh
+        user: root
+      when:
+        - backup__restic_repository is defined
+        - backup__restic_password is defined
```
I would remove this: there is no point in running the rest of the playbook if these are not set, you can just leave it to fail.

sim6 replied:
I removed it via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/666a58dc739e6f8202f75ad934e854f0ec4c544c

sim6 commented on playbooks/backup.yaml:
```diff
@@ -0,0 +1,42 @@
+        job: >
+          RESTIC_PASSWORD_FILE=/usr/local/etc/restic_password
+          RESTIC_REPOSITORY={{ backup__restic_repository }}
+          /usr/local/sbin/backup.sh
```
Thanks!

sim6 commented:
I just squashed the commits.

sim6 commented:
I rebased onto `main`.

NightTsarina commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,165 @@
+get_backup_location(){
+  # Get backup location for device name
+  local device_name=$1
+  echo "$BACKUP_FROM_DIR/$(dmsetup info -C --noheadings -o name $device_name | sed 's/--snap$//')"
```
here and in other places, it'd be good to wrap at 80 chars

sim6 replied:
I will do it.

sim6 replied:
I done it via https://github.com/Tina-pm/ansible-playbooks/pull/20/commits/828c98ae80d8b8c2bb38d290d2ddb75841e3323a

sim6 replied:
I am going to resolve it due to the approval https://github.com/Tina-pm/ansible-playbooks/pull/20#pullrequestreview-2573063558

NightTsarina reviewed (approved):
LGTM.
I am sorry I missed a few details before, so feel free to merge and address those later  if you prefer.

sim6 commented on playbooks/files/backup/backup.sh:
```diff
@@ -0,0 +1,165 @@
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
```
Thanks (but `local` works with dash).

sim6 commented on playbooks/backup.yaml:
```diff
@@ -0,0 +1,39 @@
+        dest: /usr/local/sbin/backup.sh
+        owner: root
+        group: root
+        mode: '0700'
```
It can only be executed by root due to it uses tools only accessible to root. I am missing something?

NightTsarina replied:
It is correct, but there is no security risk with giving read permissions.. like all other binaries in  `/usr/sbin`

sim6 replied:
OK, thanks.

sim6 commented:
I squashed the commits.

sim6 commented:
I squashed the commits.
